experiment/popularity_community/crawl_torrents.py

```diff
@@ -0,0 +1,214 @@
+
+        for peer in peers_ready_to_request:
+            self._logger.debug(f"Requesting torrents from peer: {peer}")
+            self.send_select(peer, metadata_type=[REGULAR_TORRENT], first=0, last=100)
+            self._logger.debug(f"Marked as polled: {peer}")
+            self.polled_peers.add(peer)
+
+    @lazy_wrapper(SelectResponsePayload)
+    async def on_remote_select_response(self, peer, response):
+        self._logger.debug(f"Received select response from {peer}")
+
+        decompressed_data = lz4.frame.decompress(response.raw_blob)
+        unpacked_data = self.mds.process_squashed_mdblob(decompressed_data)
+
+        for metadata, _ in unpacked_data:
+            if not metadata:
+                continue
+
+            self.add_torrent_to_the_list(peer, metadata)
+
+    def add_torrent_to_the_list(self, peer, metadata):
+        key = hexlify(peer.mid)
+
+        votes = int(getattr(metadata, 'votes', 0))
+        infohash = hexlify(getattr(metadata, 'infohash', ''))
+        title = getattr(metadata, 'title', '')
+
+        self._logger.info(f"Collect torrent item for {key}: {title}")
+        self.torrent_list[key].append((votes, infohash, title))
+
+    # this method has been overloaded only for changing timeout
+    def send_select(self, peer, **kwargs):
+        request = CrawlerSelectRequest(self.request_cache, hexlify(peer.mid), kwargs)
+        self.request_cache.add(request)
+        self.ez_send(peer, RemoteSelectPayload(request.number, json.dumps(kwargs).encode('utf8')))
+
+    def save(self):
+        self._logger.info(f"Save results to : {self._output_file_path}")
+
+        with open(self._output_file_path, 'w') as json_file:
+            json.dump(self.torrent_list, json_file)
+
+    def introduction_response_callback(self, peer, dist, payload):
+        """ This method adds logging behaviour to the base implementation
+        """
+        super().introduction_response_callback(peer, dist, payload)
+        self._logger.debug(f"Peer {peer} wants to introduce {payload.wan_introduction_address}")
+
+
+class Service(TinyTriblerService):
+    def __init__(self, output_file_path, timeout_in_sec, working_dir, config_path):
+        super().__init__(Service.create_config(working_dir, config_path),
+                         timeout_in_sec, working_dir, config_path)
+        self._output_file_path = output_file_path
+
+    @staticmethod
+    def create_config(working_dir, config_path):
+        config = TinyTriblerService.create_default_config(working_dir, config_path)
+
+        config.set_ipv8_enabled(True)
+        config.set_ipv8_walk_interval = IPV8_WALK_INTERVAL
+
+        config.set_chant_enabled(True)
+        return config
+
+    async def on_tribler_started(self):
+        await super().on_tribler_started()
+        session = self.session
+        peer = Peer(session.trustchain_keypair)
+
+        session.remote_query_community_community = Crawler(peer, session.ipv8.endpoint,
```

Comment:
That's an... inventive name for a variable :wink:

Reply:
Yes, misprint.